### DISCUSSION TITLE
Intégration/Tests

### DISCUSSION THREAD

**Discussion by mdeweerd** [saniho/apiEnedis#136] [General]
@saniho 
J'ai ceci dans ma trace - un appel vers 11h, c'est le seul.
Je ne sais pas si cela doit se répéter en cas "d'erreur" ou pas.

La version n'est pas celui du tag mais c'est la version 'dev' de git (donc c'est bien la 1.5.0.0. et non celle indiquée).

J'ai quelques appels qui se suivent vers 11h mais rien plus tard.
```
*******
2022-02-24 10:54:04 ERROR (SyncWorker_3) [myEnedis] header : {'Authorization': 'XXXX', 'Content-Type': 'application/json', 'call-service': 'home-assistant-myEnedis', 'ha_sensor_myenedis_version': '1.4.1.1'}
2022-02-24 10:54:04 ERROR (SyncWorker_3) [myEnedis] params : None
2022-02-24 10:54:04 ERROR (SyncWorker_3) [myEnedis] data : {"type": "daily_production", "usage_point_id": "21545875517880", "start": "2022-02-23", "end": "2022-02-24"}
2022-02-24 10:54:04 ERROR (SyncWorker_3) [myEnedis] Error JSON : {"tag":"enedis_return_ko","alert_user":true,"enedis_return":true,"support":"XXXX"}
2022-02-24 10:54:04 ERROR (SyncWorker_3) [myEnedis] ************************************************************
```

**Comment by mdeweerd** [comment]
Pour l'intégration sous "main".  Le merge automatique ne s'en sort pas trop - je pense essentiellement à cause du revert.

Ce sont essentiellement les changements de mise en forme qui gênent.

Je peux préparer une branche 'new_dev' à partir de main et j'y intègre 'dev'.
Je vérifie que se 'new_dev' peut se rejoindre à dev et il devrait pouvoir se joindre à main ensuite.

**Comment by mdeweerd** [comment]
Je vois des évolutions au niveau du test pour les 7 jours.
Je suppose que ce n'est pas encore terminé - je suis étonné qu'il passe alors que la date est forcé dans le passé et que les données sont pour ce mois-ci.

Pas de modification au code principal pour le moment - ce qui indique qu'il y a pas eu de bogue détecté ce qui est comfortant (sauf si on s'attend à en trouver) 👍

**Reply by saniho** [reply]
en fait il s'oqp il zappe le control si on passe le parametre : withControl=False
Donc pour moi, le test serait juste à mon avis . Ok pour toi ?

**Reply by mdeweerd** [reply]
Alors, le test est "juste" dans les deux sens:
- Il semble tester la fonctionnalité de base;
- Il est "juste" car il ne teste que ça alors qu'il peut tester plus.

En changéant la date et le control
```
   "patch_datetime_now", [(datetime.datetime(2022, 3, 8, 11, 22, 00))], indirect=True

...
        myE.updateLast7DaysDetails(data=dataJson, withControl=True)
```

J'ai cette erreur initialement:
```
        dateDuJour = (datetime.date.today()).strftime(_formatDateYmd)
        for x in data["meter_reading"]["interval_reading"]:
>           self._interval_length = x["interval_length"]
E           KeyError: 'interval_length'

../custom_components/apiEnedis/myDataEnedisByDayDetail.py:185: KeyError
```

qui pourraît être intéressant de traiter (et donc du coup à garder comme test).

Et en modifiant le fichier adéquat (c'était encore week1):
```
        dataFile = loadFile("Week/week2.json")
        m.post("https://enedisgateway.tech/api", text=dataFile)
```

le test passe.

En évitant de bypasser le contrôle avec 'withControl=False', on a pour avantage:
- De tester les fonctions de contrôle;
- De vérifier que l'impact de la date est bonne;
- D'éviter de modifier le code pour accomoder les tests (il faudrait supprimer le withControl=False partout, et forcer l'heure).

Du coup: le test qui plante qui déclenche le KeyError est valable aussi en tant que test car il permet de détecter que cette exception est gérée (et pas géré pour le moment).


Ce qui m'embête aussi c'est le test passe avec n'importe quelle date une fois qu'on charge week2.json.
```
@pytest.mark.parametrize(
    "patch_datetime_now", [(datetime.datetime(2022, 3,30, 11, 22, 00))], indirect=True
)
```

Alors que l'on devrait récuperer les informations pour les 7 derniers jours - pas pour les jours dans le futur ni trop loin dans le passé.

Donc dans ce sens là, je conclus que le test n'est pas assez complet / que le code retourne un résultat "invalide" par rapport à la date/heure simulée.

**Reply by saniho** [reply]
en effet, la remarque est juste
pour simplifier le suivi, j'ai cree un kanban et t'ai ajouté dessus, cela permettra de creer les idées et de les suivres pour savoir ce qui serait interessant à ajouter et si la communauté augmente
ps : j'ai pas trop de temps en ce moment pour bosser sur le projet
ps2 : je vais livrer la nouvelle release en prod( merge dev sur /main ) et lui donner un nouveau numéro de version, cela sera l'occasion de tester la modifiation automatique dans le code de la version

**Reply by mdeweerd** [reply]
Entendu - j'espère que la génération de version fonctionne - sinon je testerai dans un clone.
